fsspec/asyn.py

```diff
@@ -14,10 +14,13 @@
 
 
 async def _runner(event, coro, result, timeout=None):
+    timeout = timeout if timeout else None  # convert 0 or 0.0 to None
     if timeout is not None:
         coro = asyncio.wait_for(coro, timeout=timeout)
     try:
         result[0] = await coro
+    except asyncio.exceptions.TimeoutError:
+        result[0] = None
```

Comment:
This doesn't seem like a good idea - it would make it appear that the coroutine succeeded, but had no result. Are you relying on the timeout being raised in `sync()` before this None-result can be collected?

Reply:
I'm pretty sure you are facing a race condition. The purpose of `_runner` is to run the coroutine and put the result *OR* exception into the output. `sync()`, if it doesn't timeout, examines this value to see if it's an exception or not. So in the current for you have, in the case of the coroutine timing out, the following could happen
- `_runner` times out first, puts `None` into the output, and then `sync()` thinks `_runner` completed successfully and then returns `None`
- `sync()` times out first and raises `TimeOut`, in which case it doesn't matter what `_runner` returns.

I would argue that both these scenarios should raise Timeout

> I observed the behavior can be different on Linux/OSX and Windows.

It is true that both the low-level "event loop poll"  and threading systems are different, which might account for what you are seeing.

Reply:
I agree with you, to set result[0] to be None on asyncio.exceptions.TimeoutError in `_runner` will cause `sync` think `_runner` completed successfully and returns None.

And I also agreed that this is actually a race condition about `_runner` timeout first or `sync` timeout first.

Just for clarification, here `_runner` timeout means `asyncio.wait_for()` timeout, and `sync` timeout means for `event.wait()` timeout.

Here is a different between this two timeout. `asyncio.wait_for` will raise asyncio.exceptions.TimeoutError exception, but `threading.Event.wait` timeout will finish gracefully without any exception.

My argument is I think `_runner`, `sync` and `sync_wrapper` want to turn an async method into a sync method, so the asyncio.exceptions.TimeoutError is the exception we want to hide it inside in `sync`. So I think we still need to suppress this exception.

I agree with you that suppress the TimeoutError in `_runner` will cause `sync` think `_runner` succeeds. How about handling it in `sync`? like:
```
    event.wait(timeout)
    if isinstance(result[0], asyncio.exceptions.TimeoutError):
        #  To suppress the TimeoutError exception and returns None on timeout occurs
        return None
    elif isinstance(result[0], BaseException):
        raise result[0]
    return result[0]
```

Reply:
>  threading.Event.wait timeout will finish gracefully without any exception.

We should raise a Timeout for this case too. 

> To suppress the TimeoutError exception and returns None on timeout occurs

but this *is* an error condition, and we should raise so that the use knows it. While it is true that the underlying coroutine *might* have succeeded, the whole purpose of a timeout is that, if things take longer than stated, we should get an exception.

Reply:
Aha, okay, that's my misunderstanding. I thought the `sync_wrapper` will turn async method to sync, and hide the timeout situation.

Then user of `sync_wrapper` should expect to get a timeout exception when they call it with a timeout, right?

Then in this case, here we might should consider to raise a timeout exception when `threading.Event.wait()` timeout. And it should be the same exception as `asyncio.wait_for()` timeout, because user don't really care which timeout it is.

In this case, maybe we should not re-raise the `asyncio.exceptions.TimeoutError`. Instead of that, we can create something like a `fsspec.TimeoutError` exception class and raise it on either timeout situation.

I will change my PR to catch the timeout exception in test case and change it back in `_runner`. If you agree with my previous idea (raise `fsspec.TimeoutError` on both timeout situations, not re-use `asyncio.exceptions.TimeoutError`), please let me know and I can do that in another PR.

Reply:
Yes, I think you have it right.

> Then user of sync_wrapper should expect to get a timeout exception when they call it with a timeout, right?

If you can figure out how the user might pass a timeout via `sync_wrapper`, sure. However, we don't expect the user to know, in general, that the call they are making is backed by an coroutine for some implementations and not others. Neither can we realistically set a global default timeout, since some operations (downloading thousands of files) might be expected to take very long.

Reply:
I have changed it back, and skip the timeout occur test case for now until the `sync_wrapper` has a consistent behavior on timeout. I can solve it later in another PR. It's actually out of scope of this PR, which is just want to treat timeout=0 as timeout=None.

Please have a double check, thanks!